File: SolaceApp/Services/FirestoreAppointmentService.swift
import Foundation
import FirebaseFirestore
import SolaceCore

/// Firebase-backed implementation of `AppointmentServicing`. Appointment
/// documents store both participant IDs (for querying, same idiom as
/// `Conversation`) and their display names, denormalized at request time.
final class FirestoreAppointmentService: AppointmentServicing, @unchecked Sendable {
    private let firestore = Firestore.firestore()

    func requestAppointment(
        studentID: String,
        studentName: String,
        counselorID: String,
        counselorName: String,
        scheduledAt: Date
    ) async throws -> Appointment {
        let reference = firestore.collection("appointments").document()
        try await reference.setData([
            "studentID": studentID,
            "studentName": studentName,
            "counselorID": counselorID,
            "counselorName": counselorName,
            "participantIDs": [studentID, counselorID],
            "scheduledAt": Timestamp(date: scheduledAt),
            "status": AppointmentStatus.pending.rawValue,
            "createdAt": FieldValue.serverTimestamp()
        ])
        return Appointment(
            id: reference.documentID,
            studentID: studentID,
            studentName: studentName,
            counselorID: counselorID,
            counselorName: counselorName,
            scheduledAt: scheduledAt
        )
    }

    func observeAppointments(for userID: String) -> AsyncStream<[Appointment]> {
        AsyncStream { continuation in
            nonisolated(unsafe) let listener = firestore.collection("appointments")
                .whereField("participantIDs", arrayContains: userID)
                .order(by: "scheduledAt")
                .addSnapshotListener { snapshot, _ in
                    guard let snapshot else { return }
                    continuation.yield(snapshot.documents.compactMap(Self.appointment(from:)))
                }
            continuation.onTermination = { _ in listener.remove() }
        }
    }

    func updateStatus(appointmentID: String, status: AppointmentStatus) async throws {
        try await firestore.collection("appointments").document(appointmentID).setData([
            "status": status.rawValue
        ], merge: true)
    }

    private static func appointment(from document: QueryDocumentSnapshot) -> Appointment? {
        let data = document.data()
        guard
            let studentID = data["studentID"] as? String,
            let counselorID = data["counselorID"] as? String,
            let scheduledAt = (data["scheduledAt"] as? Timestamp)?.dateValue(),
            let status = AppointmentStatus(rawValue: data["status"] as? String ?? "")
        else { return nil }
        return Appointment(
            id: document.documentID,
            studentID: studentID,
            studentName: data["studentName"] as? String ?? "Student",
            counselorID: counselorID,
            counselorName: data["counselorName"] as? String ?? "Counselor",
            scheduledAt: scheduledAt,
            status: status,
            createdAt: (data["createdAt"] as? Timestamp)?.dateValue() ?? Date()
        )
    }
}

File: SolaceAndroid/app/src/main/java/com/jonathanalumasa/solace/service/firebase/FirestoreAppointmentService.kt
package com.jonathanalumasa.solace.service.firebase

import com.google.firebase.Timestamp
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import com.jonathanalumasa.solace.model.Appointment
import com.jonathanalumasa.solace.model.AppointmentStatus
import com.jonathanalumasa.solace.service.AppointmentService
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.tasks.await
import java.util.Date

/**
 * Firebase-backed [AppointmentService]. Appointment documents store
 * participant IDs (for querying, same idiom as conversations) alongside
 * denormalized display names.
 */
class FirestoreAppointmentService(
    private val firestore: FirebaseFirestore = FirebaseFirestore.getInstance()
) : AppointmentService {

    override suspend fun requestAppointment(
        studentID: String,
        studentName: String,
        counselorID: String,
        counselorName: String,
        scheduledAt: Date
    ): Appointment {
        val reference = firestore.collection("appointments").document()
        reference.set(
            mapOf(
                "studentID" to studentID,
                "studentName" to studentName,
                "counselorID" to counselorID,
                "counselorName" to counselorName,
                "participantIDs" to listOf(studentID, counselorID),
                "scheduledAt" to Timestamp(scheduledAt),
                "status" to AppointmentStatus.PENDING.rawValue,
                "createdAt" to FieldValue.serverTimestamp()
            )
        ).await()

        return Appointment(
            id = reference.id,
            studentID = studentID,
            studentName = studentName,
            counselorID = counselorID,
            counselorName = counselorName,
            scheduledAt = scheduledAt
        )
    }

    override fun observeAppointments(userID: String): Flow<List<Appointment>> =
        firestore.collection("appointments")
            .whereArrayContains("participantIDs", userID)
            .orderBy("scheduledAt")
            .snapshotListFlow(::appointment)

    override suspend fun updateStatus(appointmentID: String, status: AppointmentStatus) {
        firestore.collection("appointments").document(appointmentID)
            .set(mapOf("status" to status.rawValue), SetOptions.merge())
            .await()
    }

    private fun appointment(document: DocumentSnapshot): Appointment? {
        val studentID = document.getString("studentID") ?: return null
        val counselorID = document.getString("counselorID") ?: return null
        val scheduledAt = document.getDate("scheduledAt") ?: return null
        val status = AppointmentStatus.fromRaw(document.getString("status")) ?: return null
        return Appointment(
            id = document.id,
            studentID = studentID,
            studentName = document.getString("studentName") ?: "Student",
            counselorID = counselorID,
            counselorName = document.getString("counselorName") ?: "Counselor",
            scheduledAt = scheduledAt,
            status = status,
            createdAt = document.getDate("createdAt") ?: Date()
        )
    }
}
